crates/wasi-provider/src/handle.rs

```diff
@@ -46,25 +52,125 @@ impl<R: AsyncReadExt + AsyncSeekExt + Unpin> RuntimeHandle<R> {
+/// access logs
+pub struct PodHandle<R: AsyncReadExt + AsyncSeekExt + Unpin> {
+    container_handles: RwLock<HashMap<String, RuntimeHandle<R>>>,
+    // The channel for sending a stop signal to the status updater tasks
+    status_handle: JoinHandle<()>,
+    pod: Pod,
+}
+
+impl<R: AsyncReadExt + AsyncSeekExt + Unpin> PodHandle<R> {
+    /// Creates a new pod handle that manages the given map of container names
+    /// to [crate::RuntimeHandle]s. The given pod and client are used to
+    /// maintain a reference to the kubernetes object and to be able to update
+    /// the status of that object
+    pub fn new(
+        container_handles: HashMap<String, RuntimeHandle<R>>,
+        pod: Pod,
+        client: APIClient,
+    ) -> anyhow::Result<Self> {
+        let mut channel_map = StreamMap::with_capacity(container_handles.len());
+        for (name, handle) in container_handles.iter() {
+            channel_map.insert(name.clone(), handle.status().clone());
+        }
+        // TODO: This does not allow for restarting single containers because we
+        // move the stream map and lose the ability to insert a new channel for
+        // the restarted runtime. It may involve sending things to the task with
+        // a channel
+        let cloned_pod = pod.clone();
+        let status_handle = tokio::task::spawn(async move {
+            loop {
+                let status = match channel_map.next().await {
+                    Some(s) => s,
+                    // None means everything is closed, so go ahead and exit
+                    None => return,
+                };
+                debug!(
+                    "Got status update from container {}: {:#?}",
+                    status.0, status.1
+                );
+                cloned_pod
+                    .patch_status(
+                        client.clone(),
+                        Status {
+                            message: None,
+                            container_statuses: vec![status.1],
+                        },
+                    )
+                    .await;
+            }
+        });
+        Ok(PodHandle {
+            container_handles: RwLock::new(container_handles),
+            status_handle,
+            pod,
+        })
+    }
+
+    /// Write all of the output from the specified container into the given
+    /// buffer. Returns the number of bytes written to the buffer
+    pub async fn output(
+        &mut self,
+        container_name: &str,
+        buf: &mut Vec<u8>,
+    ) -> anyhow::Result<usize> {
+        let mut handles = self.container_handles.write().await;
+        let handle =
+            handles
+                .get_mut(container_name)
+                .ok_or_else(|| ProviderError::ContainerNotFound {
+                    pod_name: self.pod.name().to_owned(),
+                    container_name: container_name.to_owned(),
+                })?;
+        handle.output(buf).await
```

Comment:
The way that this is architected, we'll hold on to the write lock while `handle.output` is running. That's ok right?

Reply:
Yes, because we need the mutable lock on that handle until we are done